### DISCUSSION TITLE
Use USB CDC ACM port in ESP32S3 using Zephyr

### DISCUSSION THREAD

**Discussion by ankit1chauhan** [zephyrproject-rtos/zephyr#71325] [Q&A]
Currently, I am using a custom board having ESP32S3 that has USB connected with DP and DM (IO20 and IO19) pins. I need an ACM port that can be used to flash an executable in ESP32 via a USB interface. 

I have checked that the esp32s3_common.dtsi file and found that it does not have any interface of cdc_acm_uart. And that is why when I am trying to set the zephyr, console to cdc_acm_uart like below overlay

`/ {
	chosen {
		zephyr,console = &cdc_acm_uart;
		zephyr,shell-uart = &cdc_acm_uart;
	};	
};

it gave me the error 
**error: static assertion failed: "node /soc/uart@60038000/cdc_acm_uart is not assigned to a USB device controller"**

So is there any way to have an ACM0 port in ESP32 using Zephyr.

**Comment by shenyi97** [comment]
The CDC-ACM port of esp32s3 is assigned to the "usb_serial" dts node. I have just tried to use it as console and it works well.

**Reply by ankit1chauhan** [reply]
Thanks for your reply...

If I am not wrong, the USB_serial has a dedicated UART interface to use as a console, and that requires a USB-UART converter to use the USB-serial interface instead of directly using DP and DM pins.

As I said, I have custom hardware ready and with USB interface so it is not feasible to use USB-UART.

If this is not the case that I am thinking, can you please share the DTS interface for usb_serial and initialization of USB in source code? That will be very great.

**Reply by shenyi97** [reply]
I use a esp32s3 board from VCC-GND Studio, which connect one of its usb port to io20 and io19.
I modify the esp32s3_devkitm_esp32s3_procpu.dts by changing the chosen node and set usb_serial node status to "okay".
`		zephyr,console = &usb_serial;
		zephyr,shell-uart = &usb_serial;`
`&usb_serial{
	status = "okay";
};`

**Reply by shenyi97** [reply]
<img width="224" alt="2024-04-11 131624" src="https://github.com/zephyrproject-rtos/zephyr/assets/22340727/97dcc25d-89b0-4484-9c8e-f7e2c29afec5">
The usb port part of my board.

**Reply by shenyi97** [reply]
<img width="637" alt="2024-04-11 132334" src="https://github.com/zephyrproject-rtos/zephyr/assets/22340727/f5617dbd-e018-429d-bcfa-609c913ccc02">
From esp32s3 technical reference, the CDC-ACM part is connected to usb serial/jtag logic and then to usb phy, so usb-uart coverter is not  needed

**Reply by ankit1chauhan** [reply]
Thanks for the prompt reply.

Let me check again with your suggested way. 
Do I need to add any device initialization in my Main? Like usb_enable or any similar initialization.

I have tested with below code in my main. And the device tree overlay contains the same usb_serial configuration as you mentioned.

`/* Check the availability of console node */
BUILD_ASSERT(DT_NODE_HAS_COMPAT(DT_CHOSEN(zephyr_console), espressif_esp32_usb_serial),
	     "Console device is not ACM CDC UART device");

const struct device *const dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_console));

if (usb_enable(NULL)) `{`
    return 0;
}

// Time to settle down the DTR 
k_sleep(K_MSEC(500));

uart_line_ctrl_get(dev, UART_LINE_CTRL_DTR, &dtr);
if(dtr) {
    LOG_DBG("Unable to initialize USB console :: %d", dtr);	
}`

**Reply by shenyi97** [reply]
No, you don't need to.

**Reply by ankit1chauhan** [reply]
Hi,

I have tried your suggestion with my code and now I can get the ttyACM0 port in my USB. Thanks.

In your previous reply, you mentioned that you are able use ACM port as a console. However, I tried the same thing but I did not get any logs on the ACM port. Below is my Project configuration and main.

prj.conf

CONFIG_SERIAL=y
CONFIG_CONSOLE=y
CONFIG_LOG=y
CONFIG_LOG_MODE_MINIMAL=y
CONFIG_UART_CONSOLE=y 


main.cpp
`
while(true) {
   LOG_INF("CDC ACM device not ready"); 
   printk("CDC ACM device not ready\n");
   k_msleep(1000);
}

`

overlay file 
`
/ {
	chosen {
		zephyr,console = &usb_serial;
		zephyr,shell-uart = &usb_serial;
	};	
};

&usb_serial {
	status = "okay";
};

`

With this configurations I am not able to get the any prints on ACM0 port. Did I miss anything?

I have also tried by enabling the USB configuration macros but no change in output

CONFIG_USB_DEVICE_STACK=y
CONFIG_USB_CDC_ACM=y


Can you please suggest the way out.
Thanks.

**Reply by ankit1chauhan** [reply]
It's working now, It was the issue of the device instance that was not getting initialized in the code. I explicitly did the same and now I can see the logs and shell on the ACM port.

Thanks for the support.

**Comment by NOHsi** [comment]
hello.
I am trying to test usb cdc acm using zephyr on esp32s3 Devkitc. 
I tested it by building samples/subsys/usb/cdc_acm with the sample code, but it did not work. 
Since the author said he made it work, could you please upload the entire code?

**Reply by shenyi97** [reply]
Sample of usb/cdc_acm need a node of usb, which doesn't exist in esp32s3 recently, So it won't work.
As metioned above, cdc-acm on esp32s3 can be used like a normal uart node, which is named as "usb_serial".

**Reply by NOHsi** [reply]
Thank you for checking!
